### PR TITLE
Create integration testing post

### PR DESCRIPTION
**Pull request by nickmyyz**
https://app.asana.com/0/1199959077348208/1200596832140273

### CODE REVIEW

File: blogposts/2021/integration-testing.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+slug: integration-testing
+heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg
+published: true
+---
+
+![Unglamorous engineering graphic](https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg)
+
+# How we added integration testing to our CI pipeline
+
+My name is Joe Chen and I’m on the Core Application team here at Sourcegraph. It’s our responsibility to cover all the needs of the other teams. Some of our previous work includes licensing enforcement, background permissions syncing, and an explicit permissions API.
```
Comment (nickmyyz) [inline blogposts/2021/integration-testing.md:18]: @unknwon Could we clarify what we mean here by "all the needs"? Can we specify, like, would saying "infrastructure needs" be accurate?

**Comment by nickmyyz** [comment]
@unknwon This final draft is ready for review! One thing: are there any PRs that you've done that might worth linking to? Could be cool to show your work, if possible

File: blogposts/2021/integration-testing.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
+---
+title: 'How we added integration testing to our CI pipeline'
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:2]:
```suggestion
title: 'How we added backend integration testing to our CI pipeline'
```

File: blogposts/2021/integration-testing.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+externalTitle: 'Adding integration testing to our CI'
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:3]:
```suggestion
externalTitle: 'Adding backend integration testing to our CI'
```

File: blogposts/2021/integration-testing.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+publishDate: 2021-12-22T18:00+02:00
+tags: [blog]
+slug: integration-testing
+heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg
+published: true
+---
+
+![Unglamorous engineering graphic](https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg)
+
+# How we added integration testing to our CI pipeline
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:16]:
```suggestion
# How we added backend integration testing to our CI pipeline
```

File: blogposts/2021/integration-testing.md (hunk 1 above)
Comment (unknwon) [reply blogposts/2021/integration-testing.md:18]: >Core Application team

This team no longer exists.

```suggestion
My name is Joe Chen and I was on the Core Application team at Sourcegraph. Our responsibility was to build and maintain the infrastructure of the Sourcegraph application for other teams. Some of our previous work includes licensing enforcement, background permissions syncing, and an explicit permissions API.
```

Currently I'm on Cloud SaaS but on a different mission, so I think current tense is not accurate. Feel free to tweak my suggestion before applying!

File: blogposts/2021/integration-testing.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+My name is Joe Chen and I’m on the Core Application team here at Sourcegraph. It’s our responsibility to cover all the needs of the other teams. Some of our previous work includes licensing enforcement, background permissions syncing, and an explicit permissions API.
+
+In April of 2020, we decided to tackle integration testing. Integration testing isn’t something any single team could own because it involves code-based infrastructure. Therefore, it fell on the Core Application team.
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:20]: I don't feel the cause-and-effect makes sense in this sentence.

```suggestion
In April of 2020, we decided to tackle integration testing. Although the integration testing isn’t something any single team could own, it involves code-based infrastructure and naturally fell on the Core Application team.
```

File: blogposts/2021/integration-testing.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+The goal was to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches, the end result being a more reliable, less flaky testing suite.
+
+Integration isn’t glamorous work, but it’s necessary and, most importantly, impactful. Inspired by a previous post, H[ow not to break a search engine or: What I learned about unglamorous engineering](https://about.sourcegraph.com/blog/how-not-to-break-a-search-engine-unglamorous-engineering/), I wanted to share another piece of unglamorous engineering that, despite its lack of glamor, made a big impact.
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:24]:
```suggestion
Integration isn’t glamorous work, but it’s necessary and, most importantly, impactful. Inspired by a previous post, [How not to break a search engine or: What I learned about unglamorous engineering](https://about.sourcegraph.com/blog/how-not-to-break-a-search-engine-unglamorous-engineering/), I wanted to share another piece of unglamorous engineering that, despite its lack of glamor, made a big impact.
```

File: blogposts/2021/integration-testing.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+## Before integration testing
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:26]:
```suggestion
## Before backend integration testing
```

File: blogposts/2021/integration-testing.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+Our work started with an [RFC](https://handbook.sourcegraph.com/company-info-and-process/communication/rfcs), or Request For Comment, that you can see [here](https://docs.google.com/document/d/1LfCDPZZAkP4gFB0no-0Hb90EqMItnzv3YEb07B7WtrM/edit#heading=h.trqab8y0kufp).
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:28]:
```suggestion
Our work started with an [RFC](https://handbook.sourcegraph.com/company-info-and-process/communication/rfcs), or Request For Comment, that you can see [on our public RFCs folder](https://docs.google.com/document/d/1LfCDPZZAkP4gFB0no-0Hb90EqMItnzv3YEb07B7WtrM/edit#heading=h.trqab8y0kufp).
```

File: blogposts/2021/integration-testing.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+Before integration testing, we were relying entirely on end-to-end tests written in Javascript. These tests would boot up a virtual browser, move the mouse, and click buttons–simulating a real user using the instance.
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:30]:
```suggestion
Before having the backend integration testing, we were relying entirely on end-to-end tests written in JavaScript. These tests would boot up a virtual browser, move the mouse, and click buttons–simulating a real user using the instance.
```

File: blogposts/2021/integration-testing.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+The problem with this method is that it’s not always stable and it’s easy to either time out or miss the button. The CI pipeline ends up becoming undependable, incurring a loss of trust in the end-to-end tests. Instead, we wanted to remove the browser manipulation layer and hit the APIs directly.
```
Comment (unknwon) [inline blogposts/2021/integration-testing.md:32]:
```suggestion
The problem with this method is that it’s not always stable and it’s easy to either timeout or miss the button. The CI pipeline ends up becoming undependable, incurring a loss of trust in the end-to-end tests. Instead, we wanted to remove the browser manipulation layer and hit the APIs directly.
```

**Review by unknwon — CHANGES_REQUESTED** [review]
For some reasons, I feel this document is very unpolished... I hope we do not expect publishing this before the holiday because I think we need another two to three rounds of feedback.

**Comment by rebeccadee** [comment]
@unknwon Nick is on PTO so I was planning to take over from him. We were hoping to publish soon but of course we want you to be happy with it before that happens! I will go through your suggestions but it would also be helpful for our planning if you could give me an idea of what you're hoping to gain from 2-3 more rounds of edits?

**Comment by unknwon** [comment]
> @unknwon Nick is on PTO so I was planning to take over from him. We were hoping to publish soon but of course we want you to be happy with it before that happens! I will go through your suggestions but it would also be helpful for our planning if you could give me an idea of what you're hoping to gain from 2-3 more rounds of edits?

My last review was about one third of the overall length of the content. I stopped and submitted because I realized that I was suggesting changes to almost every single paragraph 😅 So I am wondering if the process is faster if I just make direct pushes.

File: blogposts/2021/integration-testing.md (hunk 10 above)
Comment (rebeccadee) [reply blogposts/2021/integration-testing.md:32]: I think this is a grammatical issue – "timeout" is a noun, whereas here it's used as a verb so I would have it as two words.

**Comment by rebeccadee** [comment]
@unknwon Thanks for clarifying!

> I am wondering if the process is faster if I just make direct pushes.

Feel free to do that if it's easier. If there are places where you'd like more help with rewriting I can do that, but I agree if it's small edits to make it flow better it will be quicker just to make those yourself! Thanks for your patience, we haven't worked on a post in this manner at Sourcegraph before, so your input will help us make the process smoother next time 🤝

**Comment by unknwon** [comment]
@rebeccadee I push my proofread in https://github.com/sourcegraph/about/pull/5037/commits/3b542b00523106f488421bd9c80a26fe9b6bd482. PTAL!

File: blogposts/2021/integration-testing.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
+---
+title: 'How we added backend integration testing to our CI pipeline'
+externalTitle: 'Adding backend integration testing to our CI'
+description: 'Here's the story, and the lessons learned, from our work to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches.
```
Comment (rebeccadee) [inline blogposts/2021/integration-testing.md:4]:
```suggestion
description: "Here's the story, and the lessons learned, from our work to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches."
```

File: blogposts/2021/integration-testing.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+externalDescription: 'Integration testing isn't glamorous work, but it is impactful. Here's a behind-the-scenes look at how we worked toward a more reliable, less flaky testing suite.'
```
Comment (rebeccadee) [inline blogposts/2021/integration-testing.md:5]:
```suggestion
externalDescription: "Integration testing isn't glamorous work, but it is impactful. Here's a behind-the-scenes look at how we worked toward a more reliable, less flaky testing suite."
```

**Comment by rebeccadee** [comment]
@unknwon Thank you so much for your suggestions. I'm sorry I didn't get to it before the holiday! I'm handing back over to @nickmyyz who is back from PTO too. Nick, I've made some suggestions that I think will fix the failing build (single quotation marks in the frontmatter often causes issues if there's an apostrophe in the field too!)

**Comment by nickmyyz** [comment]
@unknwon thanks for your patience with the changes on this one! I took another pass and made some small grammar/wording changes. Would you mind taking another look and letting me know how you're feeling about the article now?

**Review by unknwon — APPROVED** [review]
LGTM, thank you both!

File: blogposts/2021/integration-testing.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
+---
+title: 'How we added backend integration testing to our CI pipeline'
+externalTitle: 'Adding backend integration testing to our CI'
+description: "Here's the story, and the lessons learned, from our work to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches."
+externalDescription: "Integration testing isn't glamorous work, but it is impactful. Here's a behind-the-scenes look at how we worked toward a more reliable, less flaky testing suite."
+author: Joe Chen
+publishDate: 2021-12-22T18:00+02:00
+tags: [blog]
+slug: integration-testing
+heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg
```
Comment (rebeccadee) [inline blogposts/2021/integration-testing.md:10]: @nickmyyz Can you add this as the social image too please? (just noticed as I was sharing the preview with our designer Kate!)

**Comment by nickmyyz** [comment]
@rebeccadee I tried to publish this and some of the checks are still failing––any idea why? I'm not seeing anything wrong with the front matter.

File: blogposts/2021/integration-testing.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
+---
+title: 'How we added backend integration testing to our CI pipeline'
+externalTitle: 'Adding backend integration testing to our CI'
+description: "Here's the story, and the lessons learned, from our work to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches."
+externalDescription: "Integration testing isn't glamorous work, but it is impactful. Here's a behind-the-scenes look at how we worked toward a more reliable, less flaky testing suite."
+author: Joe Chen
+publishDate: 2021-12-22T18:00+02:00
```
Comment (rebeccadee) [inline blogposts/2021/integration-testing.md:7]: @nickmyyz I don't think this is the problem, but you should update the publish date here

**Comment by rebeccadee** [comment]
@nickmyyz I don't recognise the error message so it may not have been a problem with this PR specifically. I'm rerunning the pipeline to see if it works now!

**Comment by rebeccadee** [comment]
@nickmyyz checks all passed ✅

**Pull request merged**